075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+1. On the **1st** reconciliation (scaling up is requested): 
+     * the `podSet()` method triggers the scaling up, as described in the previous section, but the reconciliation is not blocked on waiting for the newly added pods to be ready.
+     * the `autoRebalancing()` method detects that a scaling up was requested because of `KafkaCluster.addedNodes()` returning a non empty list with the brokers' IDs to be added to the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so the state is updated as `ScalingUp` and the `add-brokers` with the brokers' IDs. Of course, it doesn't trigger any rebalance because the scaling up is going on asynchronously.
+2. From the **2nd** to the **(N-1)th** reconciliation (scaling up is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers were added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, and it has to check if it's finished or not. This can be achieved looking at the involved `StrimziPodSet`(s) and comparing `readyPods` and `pods` fields in the corresponding status. If they differ, the scaling up is still running. It should not take long time but it could go across multiple reconciliation, so this check repeats.
+3. On the **Nth** reconciliation (scaling up is done, rebalancing can start):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, but finally looking at the involved `StrimziPodSet`(s), it gets `readyPods` equals to `pods` so the scaling up is done. A new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`). The `KafkaRebalanceAssemblyOperator` takes care of handling the rebalancing.
+4. From the **(N+1)th** to the **(M-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers are added this time. 
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+5. On the **Mth** reconciliation (rebalancing id done):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen after the scaling up:
```

Comment:
Scale up is something already available today so it has its own error handling.
Of course, if scaling up doesn't end correctly the autorebalancing won't start because it's waiting for the added new nodes to be running and ready.